h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -205,23 +208,39 @@ public void onBuildFailure(final ParallelModelBuilder.ModelBuildFailure modelBui
       } finally {
         parallelSearchGridLock.unlock();
       }
+    }
+
+    @Override
+    public void postBuildSuccess(Model finishedModel, ParallelModelBuilder parallelModelBuilder) {
+      attemptBuildNextModel(parallelModelBuilder, finishedModel);
+    }
+
+    @Override
+    public void postBuildFailure(ParallelModelBuilder.ModelBuildFailure modelBuildFailure, ParallelModelBuilder parallelModelBuilder) {
       attemptBuildNextModel(parallelModelBuilder, null);
     }
 
     private void attemptBuildNextModel(final ParallelModelBuilder parallelModelBuilder, final Model previousModel) {
-      // Attempt to train next model
-      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
-      if (nextModelParams != null
-              && isThereEnoughTime()
-              && !_job.stop_requested()
-              && !_hyperSpaceWalker.stopEarly(previousModel, grid.getScoringInfos())) {
 
-        // FIXME: ModelFeeder's <MP> type parameter is hiding GridSearch's one. Ideally would have moved ModelFeeder outside but it is too coupled with GridSearch class.
-        reconcileMaxRuntime(grid._key, hyperspaceIterator, nextModelParams);
+      synchronized (hyperspaceIterator) {
+        int numberOfExpectedToCompleteModels = parallelModelBuilder.getModelInProgressCounter().get();
+        int numberOfCompletedModels = parallelModelBuilder.getModelCompletedCounter().get();
 
-        parallelModelBuilder.run(Collections.singletonList(ModelBuilder.make(nextModelParams)));
-      } else {
-        parallelModelBuilder.noMoreModels();
+        if (hyperspaceIterator.hasNext(previousModel)
+                && isThereEnoughTime()
+                && (hyperspaceIterator.max_models() == 0 || numberOfCompletedModels + numberOfExpectedToCompleteModels < hyperspaceIterator.max_models())
+                && !_job.stop_requested()
+                && (previousModel == null || !_hyperSpaceWalker.stopEarly(previousModel, grid.getScoringInfos()))) {
```

Comment:
can't the `stopEarly` method handle the null case?

Reply:
@seb-h2o this `previousModel == null` covers case when we have failed to evaluate particular model parameters and we can't make new decision about early stopping untill we get new data ( new models). Not sure if this logic could be moved inside `stopEarly`. Need to think about it.

Reply:
@seb-h2o decided to move null case inside `stopEarly`